scylla/src/policies/speculative_execution.rs

```diff
@@ -195,3 +195,132 @@ where
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    // Important to start tests with paused clock. If starting unpaused, and calling `tokio::time::pause()`, then
+    // things like `sleep` will advance the timer not fully accurately (I have no idea why), causing
+    // few ms added clock advancement at the end of the test.
+    // Starting paused is done with `#[tokio::test(flavor = "current_thread", start_paused = true)]`.
+
+    #[cfg(feature = "metrics")]
+    use std::sync::Arc;
+    use std::sync::LazyLock;
+    use std::time::Duration;
+
+    use assert_matches::assert_matches;
+
+    use crate::errors::{RequestAttemptError, RequestError};
+    #[cfg(feature = "metrics")]
+    use crate::observability::metrics::Metrics;
+    use crate::policies::speculative_execution::{
+        execute, Context, SimpleSpeculativeExecutionPolicy,
+    };
+    use crate::response::Coordinator;
+
+    static EMPTY_CONTEXT: LazyLock<Context> = LazyLock::new(|| Context {
+        #[cfg(feature = "metrics")]
+        metrics: Arc::new(Metrics::new()),
+    });
+
+    static IGNORABLE_ERROR: Option<Result<((), Coordinator), RequestError>> = Some(Err(
+        RequestError::LastAttemptError(RequestAttemptError::UnableToAllocStreamId),
+    ));
+
+    #[tokio::test(flavor = "current_thread", start_paused = true)]
+    async fn test_exhausted_plan_with_running_fibers() {
+        let policy = SimpleSpeculativeExecutionPolicy {
+            max_retry_count: 5,
+            retry_interval: Duration::from_secs(1),
+        };
+
+        let generator = {
+            // Index of the fiber, 0 for first execution.
+            let mut counter = 0;
+            move |_first: bool| {
+                let future = {
+                    let fiber_idx = counter;
+                    async move {
+                        if fiber_idx < 4 {
+                            tokio::time::sleep(Duration::from_secs(5)).await;
+                            IGNORABLE_ERROR.clone()
+                        } else if fiber_idx == 4 {
+                            None
+                        } else {
+                            panic!("Too many speculative executions - expected 4");
+                        }
+                    }
+                };
+                counter += 1;
+                future
+            }
+        };
+
+        let now = tokio::time::Instant::now();
+        let res = execute(&policy, &EMPTY_CONTEXT, generator).await;
```

Comment:
:pick: This function would be more readable if qualified by its module. The mere `execute` sounds vague.

Reply:
Is `super::execute` fine?

Reply:
Changed to `super::execute`, lmk if its fine.